aehmc/trajectory.py

```diff
@@ -37,3 +68,431 @@ def one_step(q, p, energy, energy_grad):
+            new_left_state = where_state(do_go_right, left_state, new_state)
+            new_right_state = where_state(do_go_right, new_state, right_state)
+
+            # Update the proposal If the termination criterion is reached in
+            # the subtree or if a divergence occurs we reject this subtree's
+            # proposal. We nevertheless update the sum of the logarithm of the
+            # acceptance probabilities to serve as an estimate for dual
+            # averaging.
+            updated_sum_log_p_accept = _logaddexp(proposal[3], new_proposal[3])
+            updated_proposal = (
+                proposal[0],
+                proposal[1],
+                proposal[2],
+                updated_sum_log_p_accept,
+            )
+
+            sampled_proposal = where_proposal(
+                is_diverging | is_subtree_turning,
+                updated_proposal,
+                proposal_sampler(srng, proposal, new_proposal),
+            )
+
+            # check if the trajectory is turning
+            is_turning = uturn_check_fn(
+                new_left_state[1], new_right_state[1], new_momentum_sum
+            )
+
+            do_stop_expanding = is_diverging | is_turning
+
+            return (
+                step + 1,
+                *sampled_proposal[0],
+                sampled_proposal[1],
+                sampled_proposal[2],
+                sampled_proposal[3],
+                *new_left_state,
+                *new_right_state,
+                new_momentum_sum,
+                *new_termination_state,
+            ), until(do_stop_expanding)
+
+        # results, _ = expand_once(
+        #     0,
+        #     *proposal[0],
+        #     proposal[1],
+        #     proposal[2],
+        #     proposal[3],
+        #     *left_state,
+        #     *right_state,
+        #     momentum_sum,
+        #     *termination_state,
+        # )
+
+        results, updates = aesara.scan(
+            expand_once,
+            outputs_info=(
+                0,
+                *proposal[0],
+                proposal[1],
+                proposal[2],
+                proposal[3],
+                *left_state,
+                *right_state,
+                momentum_sum,
+                *termination_state,
+            ),
+            n_steps=max_num_expansions,
+        )
+        for key, value in updates.items():
+            key.default_update = value
+
+        return results[1:], updates
+
+    return expand
+
+
+def where_state(
+    do_pick_left: bool,
+    left_state: IntegratorStateType,
+    right_state: IntegratorStateType,
+) -> IntegratorStateType:
+    """Represents a switch between two states depending on a condition."""
+    q_left, p_left, potential_energy_left, potential_energy_grad_left = left_state
+    q_right, p_right, potential_energy_right, potential_energy_grad_right = right_state
+
+    q = aet.where(do_pick_left, q_left, q_right)
+    p = aet.where(do_pick_left, p_left, p_right)
```

Comment:
Are these supposed to be vectorized conditions, or scalar conditions that choose one vector or another?  

In the tests, `do_pick_left` is a scalar and `q_left` and `q_right` are vectors (albeit vectors of length one).  If you make `q_left` and `q_right` scalars, this should work like an `if`-`else`; otherwise, if `q_left` and `q_right` are not necessarily scalars, you might be better off using `ifelse`.

Reply:
I need to use `ifelse` instead of `where`.